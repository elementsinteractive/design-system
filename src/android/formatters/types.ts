export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}
