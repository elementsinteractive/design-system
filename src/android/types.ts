export type FormatterConfig = {
  template: string
  header: string
  packageName: string
}

export type ThemeFileOptions = {
  destination: string
  packageName: string
}

export type AndroidTheme = {
  typography: ThemeFileOptions

  colors: ThemeFileOptions

  spacings: ThemeFileOptions
}

export type AndroidJsonOptions = {
  theme: AndroidTheme
}
