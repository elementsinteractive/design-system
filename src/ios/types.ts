export type FormatterConfig = {
  template: string
  header: string
  themeName: string
}
