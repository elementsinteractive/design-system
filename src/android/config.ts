import { AndroidConfig } from '../design.types'
import { TEMPLATES } from './constants'
import { AndroidJsonOptions } from './types'

export const createConfig = (input: string, options: AndroidJsonOptions) => {
  const { theme } = options

  const { colorsTemplate, spacingsTemplate, typographyTemplate } =
    TEMPLATES(theme)

  return {
    source: [input],
    platforms: {
      android: {
        transformGroup: 'tokens-android',
        files: [typographyTemplate, spacingsTemplate, colorsTemplate].map(
          template => ({
            format: template.formatter,
            destination: template.destination,
            filter: template.filter,
          }),
        ),
      },
    },
  }
}
