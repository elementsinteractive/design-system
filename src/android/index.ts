import { AndroidConfig } from '../design.types'
import { getThemeSD } from './theme'

export const setup = (androidConfig: AndroidConfig) => {
  console.log('Setting up Android', androidConfig)

  const { configPath, input } = androidConfig

  let StyleDictionary

  if (androidConfig.type == 'theme') {
    StyleDictionary = getThemeSD(configPath, input)
  }

  if (StyleDictionary === undefined) {
    throw new Error('Could not create StyleDictionary config')
  }

  console.log('Building for platform Android')

  StyleDictionary.buildPlatform('android')
}
