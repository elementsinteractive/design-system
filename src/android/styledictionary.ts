import { TRANSFORM_GROUP } from './constants'

import { TEMPLATES } from './constants'
import StyleDictionaryPackage from 'style-dictionary'
import { AndroidThemeOptions } from './types'
import { registerFormatters } from './formatter'

const { FORMATTERS, FILTERS } = TEMPLATES

export const registerSD = (options: AndroidThemeOptions) => {
  registerFilters()
  registerTransformGroup()
  registerFormatters(options)
}

export const getSDConfig = (input: string, options: AndroidThemeOptions) => {
  const config = {
    source: [input],
    platforms: {
      android: {
        transformGroup: TRANSFORM_GROUP,
        files: [
          {
            format: FORMATTERS.colorTokens,
            destination: options.colorTokensDestination,
            filter: FILTERS.colors,
          },
          {
            format: FORMATTERS.colorTokensTypes,
            destination: options.colorTokensTypesDestination,
            filter: FILTERS.colors,
          },
          {
            format: FORMATTERS.spacingsTokens,
            destination: options.spacingsTokensDestination,
            filter: FILTERS.spacings,
          },
          {
            format: FORMATTERS.spacingsTokensTypes,
            destination: options.spacingsTokensTypesDestination,
            filter: FILTERS.spacings,
          },
          {
            format: FORMATTERS.typography,
            destination: options.typographiesDestination,
            filter: FILTERS.typography,
          },
        ],
      },
    },
  }

  return StyleDictionaryPackage.extend(config)
}

export const registerTransformGroup = () => {
  StyleDictionaryPackage.registerTransformGroup({
    name: TRANSFORM_GROUP,
    transforms: ['name/ti/camel'],
  })
}

export const registerFilter = (name, type) => {
  StyleDictionaryPackage.registerFilter({
    name,
    matcher: token => {
      return token.type === type
    },
  })
}

const registerFilters = () => {
  registerFilter(TEMPLATES.FILTERS.colors, 'color')
  registerFilter(TEMPLATES.FILTERS.sizings, 'spacing')
  registerFilter(TEMPLATES.FILTERS.typography, 'typography')
  registerFilter(TEMPLATES.FILTERS.spacings, 'spacing')
}
