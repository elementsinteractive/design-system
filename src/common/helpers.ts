import fs from 'fs-extra'
import { readFileSync } from 'fs'

export const clearOutput = path => fs.emptyDirSync(path)

export const fileExists = filePath => {
  try {
    return fs.statSync(filePath).isFile()
  } catch (err) {
    return false
  }
}

export const fileToJSON = path => fs.readJsonSync(path)

export const fileToString = path => readFileSync(path, 'utf8').toString()
