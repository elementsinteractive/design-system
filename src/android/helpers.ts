import { readFileSync } from 'fs'
import { fileToJSON } from '../common/helpers'
import { AndroidJsonOptions } from './types'

export const readOptionsAsJson = (configPath: string) =>
  fileToJSON(configPath) as AndroidJsonOptions
