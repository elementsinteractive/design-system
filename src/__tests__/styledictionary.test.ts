import { getSDConfig, registerSD } from '../android/styledictionary'
import { join } from 'path'
import { AndroidThemeOptions } from '../android/types'
import { fileToString } from '../common/helpers'

const input = join(__dirname, './tokens.json')
const options: AndroidThemeOptions = {
  themeName: 'light',
  packageName: 'nl.elements.designsystem',
  colorTokensDestination: join(__dirname, '__output/ColorTokens.kt'),
  colorTokensTypesDestination: join(__dirname, '__output/ColorTokensTypes.kt'),
  spacingsTokensDestination: join(__dirname, '__output/SpacingsTokens.kt'),
  spacingsTokensTypesDestination: join(
    __dirname,
    '__output/SpacingsTokensTypes.kt',
  ),
  typographiesDestination: join(__dirname, '__output/Type.kt'),
}

describe('android styledictionary', () => {
  let StyleDictionaryExtended

  beforeAll(() => {
    registerSD(options)

    StyleDictionaryExtended = getSDConfig(input, options)
  })

  it('can build platform android', () => {
    StyleDictionaryExtended.buildPlatform('android')

    const files = [
      options.colorTokensDestination,
      options.colorTokensTypesDestination,
      options.spacingsTokensDestination,
      options.spacingsTokensTypesDestination,
      options.typographiesDestination,
    ]

    files.map(file => {
      expect(fileToString(file)).toMatchSnapshot()
    })
  })
})
