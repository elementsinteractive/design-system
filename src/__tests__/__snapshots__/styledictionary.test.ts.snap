// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`android styledictionary can build platform android 1`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package nl.elements.designsystem.light

import androidx.compose.ui.graphics.Color
import nl.elements.designsystem.ColorTokens

object LightColors : ColorTokens {
    override val primary: Color = Color(0xffffffff)
    override val background: Color = Color(0xfff4f4f4)
    override val primaryVariant: Color = Color(0xff000000)
    override val secondary: Color = Color(0xffff495b)
    override val secondaryVariant: Color = Color(0xfff4f4f4)
    override val surface: Color = Color(0xffffffff)
    override val onPrimary: Color = Color(0xffffffff)
    override val onSecondary: Color = Color(0xff000000)
    override val onBackground: Color = Color(0xff000000)
    override val onSurface: Color = Color(0xff000000)
    override val purple: Color = Color(0xff5048da)
    override val white: Color = Color(0xffffffff)
    override val pink: Color = Color(0xffff495b)
    override val black: Color = Color(0xff000000)
    override val lightGray: Color = Color(0xfff4f4f4)
}
"
`;

exports[`android styledictionary can build platform android 2`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package [object Object]

import androidx.compose.ui.graphics.Color

interface ColorTokens {
    val primary: Color
    
    val background: Color
    
    val primaryVariant: Color
    
    val secondary: Color
    
    val secondaryVariant: Color
    
    val surface: Color
    
    val onPrimary: Color
    
    val onSecondary: Color
    
    val onBackground: Color
    
    val onSurface: Color
    
    val purple: Color
    
    val white: Color
    
    val pink: Color
    
    val black: Color
    
    val lightGray: Color
    
}
"
`;

exports[`android styledictionary can build platform android 3`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package nl.elements.designsystem

object Spacings : SpacingTokens {
  override val xxxs = 4
  override val xxs = 8
  override val xs = 12
  override val s = 16
  override val m = 24
  override val l = 32
  override val xl = 48
  override val xxl = 200
}"
`;

exports[`android styledictionary can build platform android 4`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package nl.elements.designsystem

interface SpacingTokens {
    val xxxs: Int
    val xxs: Int
    val xs: Int
    val s: Int
    val m: Int
    val l: Int
    val xl: Int
    val xxl: Int
}
"
`;

exports[`android styledictionary can build platform android 5`] = `
"/**
 * ⚠️ DO NOT MODIFY ⚠️
 * this file is auto generated using elements-design-tokens
 *
 * See https://github.com/elementsinteractive/design-tokens for more info
 */
package nl.elements.designsystem

import androidx.compose.material3.Typography
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp

val bodyLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 16.sp,
    letterSpacing = (0.5).sp,
)

val bodyMedium = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (0.2).sp,
)

val bodySmall = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 12.sp,
    letterSpacing = (0.4).sp,
)

val displayLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 57.sp,
    letterSpacing = (-0.2).sp,
)

val displayMedium = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 45.sp,
)

val displaySmall = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 36.sp,
)

val headlineLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 32.sp,
)

val headlineMedium = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 28.sp,
)

val headlineSmall = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 24.sp,
)

val labelLarge = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (0.1).sp,
)

val labelMedium = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 12.sp,
    letterSpacing = (0.5).sp,
)

val labelSmall = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 11.sp,
    letterSpacing = (0.5).sp,
)

val titleLarge = TextStyle(
    fontWeight = FontWeight.Normal,
    fontSize = 22.sp,
)

val titleMedium = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 16.sp,
    letterSpacing = (0.2).sp,
)

val titleSmall = TextStyle(
    fontWeight = FontWeight.Medium,
    fontSize = 14.sp,
    letterSpacing = (0.1).sp,
)

val AppTypography = Typography(
    bodyLarge = bodyLarge,
    bodyMedium = bodyMedium,
    bodySmall = bodySmall,
    displayLarge = displayLarge,
    displayMedium = displayMedium,
    displaySmall = displaySmall,
    headlineLarge = headlineLarge,
    headlineMedium = headlineMedium,
    headlineSmall = headlineSmall,
    labelLarge = labelLarge,
    labelMedium = labelMedium,
    labelSmall = labelSmall,
    titleLarge = titleLarge,
    titleMedium = titleMedium,
    titleSmall = titleSmall,
)"
`;
